### DISCUSSION TITLE
Memory consumption considerations

### DISCUSSION THREAD

**Discussion by fdhsdrdark** [SixLabors/ImageSharp#1738] [Q&A] [answered]
Hello,

First time using ImageSharp and run into some increased memory issues.
Using the latest version for doing a simple Resize transformation, using the Mutate method.
Had a look on the Memory Managment section on the documentation and enabled ModeratePooling.
Tried to clean up and dispose all created objects during the operation without success. 
I have hosted my API to a poor specs server (just 1 GB shared RAM memory) but that's not seem to be the issue since even in my local machine on every image request the memory usage is drastically increased. 
In order to overcome the issue, I manually triggered Garbagge Collection and that seems to resolve temporarilly the issue. I share my code if someone can shed some light.

```c#
public byte[] GetImage(string imagePath, int? width = null, int? height = null)
		{
			try
			{
				byte[] photoBytes = File.ReadAllBytes(imagePath);
				MemoryStream inStream = new MemoryStream(photoBytes, 0, photoBytes.Length, true, true);
				MemoryStream outStream = new MemoryStream();
				if (width != null && height != null)
				{
					IImageFormat format;
					using (Image image = Image.Load(inStream, out format))
					{
						if ((int)width > (int)height)
						{
							image.Mutate(x => x.Resize(new Size { Width = (int)width, Height = 0 }).BackgroundColor(Color.FromRgb(27, 49, 71)));
							image.Save(outStream, format);
						}
						else
						{
							image.Mutate(x => x.Resize(new Size { Width = 0, Height = (int)height }).BackgroundColor(Color.FromRgb(27, 49, 71)));
							image.Save(outStream, format);
						}
					}
				}
				else
				{
					IImageFormat format;
					using (Image image = Image.Load(inStream, out format))
					{
						image.Mutate(x => x.BackgroundColor(Color.FromRgb(27, 49, 71)));
						image.Save(outStream, format);
					}
				}

				Array.Clear(photoBytes, 0, photoBytes.Length);

				ClearMemoryStream(inStream);
				inStream.Dispose();

				byte[] result = outStream.ToArray();

				ClearMemoryStream(outStream);
				outStream.Dispose();

				return result;
			}
			catch (Exception ex)
			{
				this.Logger.LogError(ex, ex.Message);
				throw;
			}
			finally
			{
				GC.Collect();
			}
		}
```

Thanks very much.

**Comment by antonfirsov** [comment, marked as answer]
What are your typical image sizes?

Why all the `MemoryStream`-s and the `byte[]` ? You can do `Image.Load(filePath)` / `Image.Load(fileStream)` as well as `image.Save(filePath, format)` and `image.Save(fileStream, format)`.

> Had a look on the Memory Managment section on the documentation and enabled ModeratePooling.
> [...]
> on every image request the memory usage is drastically increased.
> I manually triggered Garbagge Collection and that seems to resolve temporarilly the issue.

With moderate pooling, `Image` is more likely to allocate arrays which won't be cleaned up until GC triggers. I recommend to keep the default pooling, and do some stress test of your app on the server. Normally GC should clean the memory as you go (without manually running it).

If you are working with Jpegs, I also recommend to get the latest packages from our [MyGet feed](https://www.myget.org/feed/sixlabors/package/nuget/SixLabors.ImageSharp) since it includes #1694, which is a major improvement. Our next NuGet release will also include #1730.

**Reply by fdhsdrdark** [reply]
Thank you for getting back on this.

My largest image is about 1MB while most images are about 350KB.
Thanks for the overloads advice - going to change the impementation, indeed no need for the MemoryStreams. 

Also,since I am using Jpegs looking forward to the next Nuget release for all improvements! 👍 👍
